### PR TITLE
Meta data tracking

### PR DESCRIPTION
**Pull request by Jurwen**
* Update: The patch to implicit function labels are tracked through tagging a `func_index` at faces of each patch.
* Update: Adjusted the IO structure to output the patch to implicit label. The output file, `mesh.json`, now has the new result called `patch_label`.

### CODE REVIEW

File: src/mesh_connectivity.cpp [hunk 1]
```diff
@@ -57,6 +57,92 @@ void compute_mesh_edges(const std::vector<PolygonFace>& mesh_faces,
...
+        if (!visisted_face[i]) {
+            // new patch
+            //std::cout << "new patch: " << std::endl;
+            patches.emplace_back();
+            auto& patch = patches.back();
+            std::queue<size_t> Q;
+            Q.push(i);
+            patch.emplace_back(i);
+            visisted_face[i] = true;
+            absl::flat_hash_map<size_t, size_t> func_to_num; //adding a hash map that finds the number of vertices that are associated to each function
```
Comment (qnzhou) [inline src/mesh_connectivity.cpp:76]: Is this information (number of vertices for each function within a patch) necessary? Is this this purely for a correctness check? If so, the logic should be moved to unit test.

**Review by qnzhou — CHANGES_REQUESTED** [review]
Thanks Owen! Looking good for the first PR!  Here is my change requests:
1. We need a unit test to check the correctness of the code.
2. Avoid `exit()`, and use exception with some explanation of the reason.
3. Avoid comparing `size_t` value with negative numbers even if it works on some compilers.
4. Please add documentation to the header file to explain what the function does.

File: src/mesh_connectivity.cpp (hunk 1 above)
Comment (qnzhou) [reply src/mesh_connectivity.cpp:76]: Thanks! I think we should try to keep the computation code separate from validation code. Especially when validation is not cheap.

If this logical cannot be moved to unit tests, we can refactor the check out into its own function, say `check_patch_label(...)` that return true only when the check passed. This way, we can `assert(check_patch_label(...))` in the main body of the code, and the check will only be executed in debug mode.

**Comment by Jurwen** [comment]
Some changes are made accordingly: 

* Unit Test:
    * Added the patch labels' tests to all the unit tests.
    * Adjusted the input stream of implicit functions of unit tests accordingly to the latest change in implicit function package.
* Patch label validation:
    * The patch label computation is done through checking the face labels, and the validation is done through checking the vertex labels. The validation is done by calling `check_patch_label` without the assert function as the assert function will abort the program. Now it only throws an exception and error message when the validation failed. 
    * Minor coding style fixes: avoid `exit()` and comparing `size_t` to negatives
* Added header file illustration to edited portion.

**Comment by qnzhou** [comment]
> The validation is done by calling check_patch_label without the assert function as the assert function will abort the program.

I think we should use assertion in this case for the following reasons:
* This check is purely optional. If our code is correct, the check should never fail.
* This check involves some non-trivial computation and is inside of a loop. So, it will add some computation cost that can be avoided.
* Using assertion allow us to only run the check in debug build, and it does not impact the performance in release build.
* Assertion failure translates in debug build is the desired behavior. If the code is correct, this assertion should alway pass.

**Review by qnzhou — APPROVED** [review]
Thanks Owen! Looks good to me!

**Pull request merged**